### ISSUE TITLE
"Illegal instantiation for type variable T" on goog.array.sort

### ISSUE THREAD

**Issue by myphysicslab** [google/closure-compiler#991]
Using NTI (new type inference) get this warning:
```
src/lab/model/CollisionAdvance.js:607: WARNING - Illegal instantiation for type variable T.
You can only specify one type. Found [myphysicslab.lab.model.Collision, T].
  goog.array.sort(this.collisions_, function(c1, c2) {
  ^
```
on this code:
```javascript
  // sort collisions by estimated collision time, earliest first
  goog.array.sort(this.collisions_, function(c1, c2) {
    // sort NaN at back
    if (isNaN(c1.estimate))
      return 1;
    else if (isNaN(c2.estimate))
      return -1;
    else if (c1.estimate < c2.estimate)
      return -1;
    else if (c1.estimate > c2.estimate)
      return 1;
    else
      return 0;
    });
```
where `this.collisions_` is defined in the constructor
```javascript
* @param {!myphysicslab.lab.model.CollisionSim} sim
* @param {!myphysicslab.lab.model.DiffEqSolver=} diffEqSolver_opt
* @constructor
* @struct
* @implements {myphysicslab.lab.model.AdvanceStrategy}
*/
myphysicslab.lab.model.CollisionAdvance = function(sim, diffEqSolver_opt) {
  // code omitted...
  /** the current set of collisions; includes joints, contacts, imminent collisions
  * @type {!Array.<!myphysicslab.lab.model.Collision>}
  * @private
  */
  this.collisions_ = new Array();
```
I don't understand this warning: what is type variable T here?  How am I supplying two types? This is another example of issue #971.

(Note to self: this is while compiling PileTest).

**Comment by myphysicslab** [comment]
Here is an update to the above code.  (I was referencing an undefined property `estimate` previously).

```
src/lab/model/CollisionAdvance.js:607: WARNING - Illegal instantiation for type variable T.
You can only specify one type. Found [myphysicslab.lab.model.Collision, T].
  goog.array.sort(this.collisions_, function(c1, c2) {
  ^

src/lab/model/CollisionAdvance.js:608: WARNING - Cannot access property getEstimatedTime
of non-object type T.
    var est1 = c1.getEstimatedTime();
               ^

src/lab/model/CollisionAdvance.js:609: WARNING - Cannot access property getEstimatedTime
of non-object type T.
    var est2 = c2.getEstimatedTime();
               ^
```
Here is the updated code:
```javascript
  // sort collisions by estimated collision time, earliest first
  goog.array.sort(this.collisions_, function(c1, c2) {
    var est1 = c1.getEstimatedTime();
    var est2 = c2.getEstimatedTime();
    // sort NaN at back
    if (isNaN(est1))
      return 1;
    else if (isNaN(est2))
      return -1;
    else if (est1 < est2)
      return -1;
    else if (est1 > est2)
      return 1;
    else
      return 0;
    });
```
Here is some of the `Collision` interface:
```javascript
/**
@interface
*/
myphysicslab.lab.model.Collision = function() {};

/**
* @alias myphysicslab.lab.model.Collision
* @ignore
*/
var Collision = myphysicslab.lab.model.Collision;

/** Returns the estimated time when this collision has `getDistance() == 0`,
or {NaN} if unknown.
@method
@return {number} the estimated time when this collision has `getDistance() == 0`,
or {NaN} if unknown
*/
Collision.prototype.getEstimatedTime;
```

**Comment by myphysicslab** [comment]
All the errors disappear with type hints on the function passed to goog.array.sort:

```javascript
  // sort collisions by estimated collision time, earliest first
  goog.array.sort(this.collisions_,
    function(/** !myphysicslab.lab.model.Collision*/c1,
            /** !myphysicslab.lab.model.Collision*/c2) {
    var est1 = c1.getEstimatedTime();
    var est2 = c2.getEstimatedTime();
    // sort NaN at back
    if (isNaN(est1))
      return 1;
    else if (isNaN(est2))
      return -1;
    else if (est1 < est2)
      return -1;
    else if (est1 > est2)
      return 1;
    else
      return 0;
    });
```

**Comment by dimvar** [comment]
This is the same bug as #855. A fix for it should be coming up today.

**Comment by myphysicslab** [comment]
Confirmed fixed in Version: v20150505-104-g0a435e9
